lib/posix-fdtab/fdtab.c

```
@@ -432,6 +437,63 @@ static void fdtab_thread_term(struct uk_thread *child)
 
 UK_THREAD_INIT(fdtab_thread_init, fdtab_thread_term);
 
+#if CONFIG_LIBPOSIX_PROCESS_CLONE
+
+/**
+ * Duplicate the current fdtab and return a pointer to the copy.
+ *
+ * NOTE: This is a basic implementation that does not guarantee atomicity of the
+ * clone operation; we optimistically assume this will not break anything.
+ * Please revisit if this turns out to be false.
+ */
+static struct uk_fdtab *fdtab_duplicate(struct uk_fdtab *tab)
+{
+	struct uk_fdtab *ret = uk_malloc(tab->alloc, sizeof(*ret));
+
+	if (likely(ret)) {
+		ret->alloc = tab->alloc;
+		ret->fmap = (struct uk_fmap){
+			.bmap = {
+				.size = UK_FDTAB_SIZE,
+				.bitmap = (unsigned long *)ret->_bmap
+			},
+			.map = ret->_fdmap
+		};
+		uk_refcount_init(&ret->refcnt, 1);
+		for (int i = 0; i < UK_FDTAB_SIZE; i++) {
+			struct fdval v = _fdtab_get(tab, i);
+			const void *entry = NULL;
+
+			if (v.p)
+				entry = fdtab_encode(v.p, v.flags);
+			uk_fmap_set(&ret->fmap, i, entry);
+		}
+	}
+	return ret;
+}
+
+static int fdtab_clone(const struct clone_args *cl_args,
+		       size_t cl_args_len __unused,
+		       struct uk_thread *child,
+		       struct uk_thread *parent __unused)
+{
+	if (!(cl_args->flags & CLONE_FILES)) {
+		struct uk_fdtab *tab = uk_thread_uktls_var(child, active_fdtab);
+		struct uk_fdtab *newtab = fdtab_duplicate(tab);
+
+		if (unlikely(!newtab))
+			return -ENOMEM;
```

Very minor: In case `fdtab_duplicate()` fails for additional reasons in the future, instead of making assumptions on  the caller's side, consider if it would be preferable to update `fdtab_duplicate()` to return error pointer that the caller can directly decode and return. Alternatively you can consider adding a line in the description "Returns pointer on success or NULL if out of memory".

---

I see your point, though I prefer that functions who have a single possible failure state by design do not return error codes, but let the caller in higher levels interpret their failure. Returning errptrs is particularly tricky because then the caller must worry _both_ about errptrs _and_ NULLs (because NULL is _not_ an error pointer as it doesn't map to a negative errno).

`fdtab_duplicate()` falls into this single fail state group, as it's a glorified malloc + memcpy by design (and the latter cannot fail recoverably); `fdtab_clone` is the higher-level "public-facing" function that is expected to conform to the errno API, and there is IMO the best place to encode and return errors.
I'll update description to make its return values clear.

---

Up to you, especially as its not a public API it's not that critical. Just sharing my 2c.